File: c++/verticalOrder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        map<int, vector<int>> m;
        int off = 0;
        // need use BFS, other wise cannot make sure the upper levle node will appear fierst
        queue<pair<TreeNode *, int>> q;        
        vector<vector<int>> output;
        
        q.push(make_pair(root, 0));
        
        while (!q.empty()){
            TreeNode * curr = q.front().first;
            int pos = q.front().second;
            q.pop();
            
            if (curr == nullptr){
                continue;
            }
            
            m[pos].push_back(curr->val);
            q.push(make_pair(curr->left, pos-1));
            q.push(make_pair(curr->right, pos+1));
        }
                    
        for(auto v : m){
            output.push_back(v.second);
        }
        return output;
    }
};

File: python/verticalOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        queue = collections.deque([])
        result = []
        order = collections.defaultdict(list)
        if root is None:
            return result
        queue.append((root, 0))
        while(queue):
            node, level = queue.popleft()
            if node:
                order[level].append(node.val)
                queue.append((node.left, level - 1))
                queue.append((node.right, level + 1))

        for i in sorted(order.keys()):
            result.append(order[i])

        return result
